--- mosh-1.2.5/Makefile.am
+++ mosh-1.2.5/Makefile.am
@@ -36,12 +36,7 @@
 
 VERSION.stamp:
 	@echo @PACKAGE_STRING@ > VERSION.dist
-	@set -e; if git describe --dirty --always > VERSION.git 2>&1 && \
-		[ -z `git rev-parse --show-prefix` ]; then \
-		if ! diff -q VERSION.git VERSION.stamp > /dev/null 2>&1; then \
-			mv -f VERSION.git VERSION.stamp; \
-		fi; \
-	elif ! diff -q VERSION.dist VERSION.stamp > /dev/null 2>&1; then \
+	@set -e; if ! diff -q VERSION.dist VERSION.stamp > /dev/null 2>&1; then \
 		mv -f VERSION.dist VERSION.stamp; \
 	fi
 	@rm -f VERSION.dist VERSION.git
