diff --git a/Swiften/SConscript b/Swiften/SConscript
index 379576d..7096a9c 100644
--- a/Swiften/SConscript
+++ b/Swiften/SConscript
@@ -425,8 +425,11 @@ if env["SCONS_STAGE"] == "build" :
 
 	# Install swiften
 	if swiften_env.get("SWIFTEN_INSTALLDIR", "") :
-		swiften_env.Install(os.path.join(swiften_env["SWIFTEN_INSTALLDIR"], "lib"), swiften_lib)
+		libdir = os.path.join(swiften_env["SWIFTEN_INSTALLDIR"], "lib")
+		if swiften_env.get("SWIFTEN_LIBDIR", "") :
+			libdir = swiften_env["SWIFTEN_LIBDIR"]
+		swiften_env.Install(libdir, swiften_lib)
 		for alias in myenv["SWIFTEN_LIBRARY_ALIASES"] :
-			myenv.Command(myenv.File(os.path.join(swiften_env["SWIFTEN_INSTALLDIR"], "lib", alias)), [env.Value(swiften_lib[0].name), swiften_lib[0]], symlink)
+			myenv.Command(myenv.File(os.path.join(libdir, alias)), [env.Value(swiften_lib[0].name), swiften_lib[0]], symlink)
 		for include in swiften_includes :
 			swiften_env.Install(os.path.join(swiften_env["SWIFTEN_INSTALLDIR"], "include", os.path.dirname(include)), "#/" + include)

--- a/BuildTools/SCons/SConscript.boot
+++ b/BuildTools/SCons/SConscript.boot
@@ -220,7 +220,7 @@ env["TEST_CREATE_LIBRARIES"] = "create_test_libraries" in ARGUMENTS
 
 # Packaging
 env["DIST"] = "dist" in ARGUMENTS or env.GetOption("clean")
-for path in ["SWIFT_INSTALLDIR", "SWIFTEN_INSTALLDIR"] :
+for path in ["SWIFT_INSTALLDIR", "SWIFTEN_INSTALLDIR", "SWIFTEN_LIBDIR"] :
 	if ARGUMENTS.get(path, "") :
 		if os.path.isabs(ARGUMENTS[path]) :
 			env[path] = Dir(ARGUMENTS[path]).abspath
