--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -92,6 +92,7 @@
 init_target(tg_owt)
 
 if (TG_OWT_USE_PIPEWIRE)
+    link_glib(tg_owt)
     if (TG_OWT_DLOPEN_PIPEWIRE)
         include(cmake/generate_stubs.cmake)
         set(stubs_loc ${webrtc_loc}/modules/desktop_capture/linux/wayland)
@@ -2458,6 +2458,13 @@
     )
 endif()
 
+if (NOT TG_OWT_USE_PIPEWIRE)
+    remove_target_sources(tg_owt ${webrtc_loc}
+        modules/desktop_capture/linux/base_capturer_pipewire.cc
+        modules/desktop_capture/linux/base_capturer_pipewire.h
+    )
+endif()
+
 if (NOT TG_OWT_BUILD_AUDIO_BACKENDS)
     remove_target_sources(tg_owt ${webrtc_loc}
         modules/audio_device/include/audio_device_factory.cc
