--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -28,7 +28,12 @@
 cmake_dependent_option(TG_OWT_USE_PIPEWIRE "Use pipewire for desktop capture." ON "UNIX; NOT APPLE" OFF)
 cmake_dependent_option(TG_OWT_DLOPEN_PIPEWIRE "dlopen pipewire for desktop capture." ${not_packaged_build} TG_OWT_USE_PIPEWIRE OFF)
 option(TG_OWT_BUILD_AUDIO_BACKENDS "Build webrtc audio backends." OFF)
+option(TG_OWT_BUILD_PULSE_BACKEND "Build pulseaudio audio backend on non-Apple Unix." ON)
+
+if (NOT TG_OWT_BUILD_AUDIO_BACKENDS)
+    set(TG_OWT_BUILD_PULSE_BACKEND OFF)
+endif()
 
 if (BUILD_SHARED_LIBS)
     set(CMAKE_POSITION_INDEPENDENT_CODE ON)
     message(WARNING
@@ -162,9 +167,11 @@
 link_libvpx(tg_owt)
 
 if (TG_OWT_BUILD_AUDIO_BACKENDS AND (UNIX AND NOT APPLE))
-    link_libalsa(tg_owt)
-    link_libpulse(tg_owt)
     link_dl(tg_owt)
+    link_libalsa(tg_owt)
+    if (TG_OWT_BUILD_PULSE_BACKEND)
+        link_libpulse(tg_owt)
+    endif()
 endif()
 
 if (NOT WIN32 AND NOT APPLE)
@@ -2218,12 +2225,6 @@
         modules/audio_device/linux/audio_mixer_manager_alsa_linux.h
         modules/audio_device/linux/latebindingsymboltable_linux.cc
         modules/audio_device/linux/latebindingsymboltable_linux.h
-        modules/audio_device/linux/audio_device_pulse_linux.cc
-        modules/audio_device/linux/audio_device_pulse_linux.h
-        modules/audio_device/linux/audio_mixer_manager_pulse_linux.cc
-        modules/audio_device/linux/audio_mixer_manager_pulse_linux.h
-        modules/audio_device/linux/pulseaudiosymboltable_linux.cc
-        modules/audio_device/linux/pulseaudiosymboltable_linux.h
         modules/audio_device/mac/audio_device_mac.cc
         modules/audio_device/mac/audio_device_mac.h
         modules/audio_device/mac/audio_mixer_manager_mac.cc
@@ -2243,6 +2244,17 @@
     )
 endif()
 
+if (NOT TG_OWT_BUILD_PULSE_BACKEND)
+    remove_target_sources(tg_owt ${webrtc_loc}
+        modules/audio_device/linux/audio_device_pulse_linux.cc
+        modules/audio_device/linux/audio_device_pulse_linux.h
+        modules/audio_device/linux/audio_mixer_manager_pulse_linux.cc
+        modules/audio_device/linux/audio_mixer_manager_pulse_linux.h
+        modules/audio_device/linux/pulseaudiosymboltable_linux.cc
+        modules/audio_device/linux/pulseaudiosymboltable_linux.h
+    )
+endif()
+
 if (NOT WIN32)
     remove_target_sources(tg_owt ${webrtc_loc}
         rtc_base/win32.cc
--- a/cmake/libwebrtcbuild.cmake
+++ b/cmake/libwebrtcbuild.cmake
@@ -50,8 +50,13 @@
     target_compile_definitions(libwebrtcbuild
     INTERFACE
         WEBRTC_ENABLE_LINUX_ALSA
-        WEBRTC_ENABLE_LINUX_PULSE
     )
+    if (TG_OWT_BUILD_PULSE_BACKEND)
+        target_compile_definitions(libwebrtcbuild
+        INTERFACE
+            WEBRTC_ENABLE_LINUX_PULSE
+        )
+    endif()
 endif()
 
 if (WIN32)
