--- a/src/modules/desktop_capture/desktop_capture_options.h
+++ b/modules/desktop_capture/desktop_capture_options.h
@@ -197,6 +197,9 @@ #if defined(WEBRTC_USE_PIPEWIRE)
 
   void set_height(uint32_t height) { height_ = height; }
   uint32_t get_height() const { return height_; }
+#else
+  bool allow_pipewire() const { return false; }
+  void set_allow_pipewire(bool allow) { }
 #endif
 
  private:

