diff --git a/src/ngx_http_lua_initworkerby.c b/src/ngx_http_lua_initworkerby.c
index c19cffd..acd9ba6 100644
--- a/src/ngx_http_lua_initworkerby.c
+++ b/src/ngx_http_lua_initworkerby.c
@@ -225,11 +225,13 @@ ngx_http_lua_init_worker(ngx_cycle_t *cycle)
 
     clcf = ngx_http_get_module_loc_conf(r, ngx_http_core_module);
 
-#if defined(nginx_version) && nginx_version >= 1003014
+#if defined(nginx_version) && nginx_version >= 100900
 
-    ngx_http_set_connection_log(r->connection, clcf->error_log);
+	ngx_set_connection_log(r->connection, clcf->error_log);
 
-#else
+#elif defined(nginx_version) && nginx_version < 100900 && nginx_version >= 1003014
+	ngx_http_set_connection_log(r->connection, clcf->error_log);
+#else
 
     c->log->file = clcf->error_log->file;
 
diff --git a/src/ngx_http_lua_timer.c b/src/ngx_http_lua_timer.c
index 48024a9..b95fc09 100644
--- a/src/ngx_http_lua_timer.c
+++ b/src/ngx_http_lua_timer.c
@@ -348,11 +348,13 @@ ngx_http_lua_timer_handler(ngx_event_t *ev)
 
     clcf = ngx_http_get_module_loc_conf(r, ngx_http_core_module);
 
-#if defined(nginx_version) && nginx_version >= 1003014
+#if defined(nginx_version) && nginx_version >= 100900
 
-    ngx_http_set_connection_log(r->connection, clcf->error_log);
+	ngx_set_connection_log(r->connection, clcf->error_log);
 
-#else
+#elif defined(nginx_version) && nginx_version < 100900 && nginx_version >= 1003014
+	ngx_http_set_connection_log(r->connection, clcf->error_log);
+#else
 
     c->log->file = clcf->error_log->file;
 
