--- build/basics.rb     2013-03-05 16:55:02.000000000 +0700
+++ build/basics.rb     2013-04-03 12:42:46.012231095 +0700
@@ -168,6 +168,7 @@
 	flags << " -fno-omit-frame-pointers" if USE_ASAN
 	flags << " -DPASSENGER_DISABLE_THREAD_LOCAL_STORAGE" if !boolean_option('PASSENGER_THREAD_LOCAL_STORAGE', true)
 end
+EXTRA_CXXFLAGS << " #{ENV['CXXFLAGS']}"

 # Extra linker flags that should always be passed to the linker.
 # These should be included first in the command string, before anything else.
--- build/packaging.rb	2013-07-09 12:10:02.000000000 +0400
+++ build/packaging.rb	2013-07-10 23:59:46.802236526 +0400
@@ -322,34 +322,29 @@
 end
 
 desc "Create a fakeroot, useful for building native packages"
-task :fakeroot => [:apache2, :nginx, :doc] do
+task :fakeroot => [:nginx, :doc] do
 	require 'rbconfig'
 	require 'fileutils'
 	include RbConfig
-	fakeroot = "pkg/fakeroot"
+	fakeroot = ENV['D']
 	
 	# We don't use CONFIG['archdir'] and the like because we want
 	# the files to be installed to /usr, and the Ruby interpreter
 	# on the packaging machine might be in /usr/local.
-	fake_rubylibdir = "#{fakeroot}/usr/lib/ruby/vendor_ruby"
-	fake_nodelibdir = "#{fakeroot}/usr/share/#{GLOBAL_NAMESPACE_DIRNAME}/node"
-	fake_libdir = "#{fakeroot}/usr/lib/#{GLOBAL_NAMESPACE_DIRNAME}"
-	fake_native_support_dir = "#{fakeroot}/usr/lib/ruby/#{CONFIG['ruby_version']}/#{CONFIG['arch']}"
-	fake_agents_dir = "#{fakeroot}/usr/lib/#{GLOBAL_NAMESPACE_DIRNAME}/agents"
-	fake_helper_scripts_dir = "#{fakeroot}/usr/share/#{GLOBAL_NAMESPACE_DIRNAME}/helper-scripts"
-	fake_resources_dir = "#{fakeroot}/usr/share/#{GLOBAL_NAMESPACE_DIRNAME}"
-	fake_include_dir = "#{fakeroot}/usr/share/#{GLOBAL_NAMESPACE_DIRNAME}/include"
-	fake_docdir = "#{fakeroot}/usr/share/doc/#{GLOBAL_NAMESPACE_DIRNAME}"
-	fake_bindir = "#{fakeroot}/usr/bin"
-	fake_sbindir = "#{fakeroot}/usr/sbin"
-	fake_apache2_module_dir = "#{fakeroot}/usr/lib/apache2/modules"
-	fake_apache2_module = "#{fake_apache2_module_dir}/mod_passenger.so"
-	fake_ruby_extension_source_dir = "#{fakeroot}/usr/share/#{GLOBAL_NAMESPACE_DIRNAME}/ruby_extension_source"
-	fake_nginx_module_source_dir = "#{fakeroot}/usr/share/#{GLOBAL_NAMESPACE_DIRNAME}/ngx_http_passenger_module"
-	
-	sh "rm -rf #{fakeroot}"
-	sh "mkdir -p #{fakeroot}"
-	
+	fake_rubylibdir = "#{fakeroot}/#{CONFIG['sitedir']}/#{CONFIG['ruby_version']}/#{CONFIG['arch']}"
+	fake_nodelibdir = "#{fakeroot}/usr/libexec/#{GLOBAL_NAMESPACE_DIRNAME}/node"
+	fake_libdir = "#{fakeroot}/" + CONFIG['sitedir'].gsub(/(\/.*?\/.*?)\/.*/,"\\1") + "/#{GLOBAL_NAMESPACE_DIRNAME}"
+	fake_native_support_dir = "#{fakeroot}/#{CONFIG['sitedir']}/#{CONFIG['ruby_version']}/#{CONFIG['arch']}"
+	fake_agents_dir = "#{fakeroot}/usr/libexec/#{GLOBAL_NAMESPACE_DIRNAME}/agents"
+	fake_helper_scripts_dir = "#{fakeroot}/usr/libexec/#{GLOBAL_NAMESPACE_DIRNAME}/helper-scripts"
+	fake_resources_dir = "#{fakeroot}/usr/libexec/#{GLOBAL_NAMESPACE_DIRNAME}/resources"
+	fake_include_dir = "#{fakeroot}/usr/include/#{GLOBAL_NAMESPACE_DIRNAME}"
+	fake_docdir = "#{fakeroot}/usr/share/doc/"+ENV['P']+"/#{GLOBAL_NAMESPACE_DIRNAME}"
+	fake_bindir = "#{fakeroot}/usr/bin"
+	fake_sbindir = "#{fakeroot}/usr/sbin"
+	fake_ruby_extension_source_dir = "#{fakeroot}/usr/src/#{GLOBAL_NAMESPACE_DIRNAME}/ruby_extension_source"
+	fake_nginx_module_source_dir = "#{fakeroot}/usr/src/#{GLOBAL_NAMESPACE_DIRNAME}/ngx_http_passenger_module"
+
 	# Ruby sources
 	sh "mkdir -p #{fake_rubylibdir}"
 	sh "cp #{PhusionPassenger.ruby_libdir}/phusion_passenger.rb #{fake_rubylibdir}/"
@@ -430,10 +430,6 @@
 		end
 	end
 	
-	# Apache 2 module
-	sh "mkdir -p #{fake_apache2_module_dir}"
-	sh "cp #{APACHE2_MODULE} #{fake_apache2_module_dir}/"
-
 	# Ruby extension sources
 	sh "mkdir -p #{fake_ruby_extension_source_dir}"
 	sh "cp -R #{PhusionPassenger.ruby_extension_source_dir}/* #{fake_ruby_extension_source_dir}"
@@ -441,19 +441,19 @@
 	puts "Creating #{fake_rubylibdir}/phusion_passenger/locations.ini"
 	File.open("#{fake_rubylibdir}/phusion_passenger/locations.ini", "w") do |f|
 		f.puts "[locations]"
-		f.puts "natively_packaged=true"
-		f.puts "bin_dir=/usr/bin"
-		f.puts "agents_dir=/usr/lib/#{GLOBAL_NAMESPACE_DIRNAME}/agents"
-		f.puts "lib_dir=/usr/lib/#{GLOBAL_NAMESPACE_DIRNAME}"
-		f.puts "helper_scripts_dir=/usr/share/#{GLOBAL_NAMESPACE_DIRNAME}/helper-scripts"
-		f.puts "resources_dir=/usr/share/#{GLOBAL_NAMESPACE_DIRNAME}"
-		f.puts "include_dir=/usr/share/#{GLOBAL_NAMESPACE_DIRNAME}/include"
-		f.puts "doc_dir=/usr/share/doc/#{GLOBAL_NAMESPACE_DIRNAME}"
-		f.puts "ruby_libdir=/usr/lib/ruby/vendor_ruby"
-		f.puts "node_libdir=/usr/share/#{GLOBAL_NAMESPACE_DIRNAME}/node"
-		f.puts "apache2_module_path=/usr/lib/apache2/modules/mod_passenger.so"
-		f.puts "ruby_extension_source_dir=/usr/share/#{GLOBAL_NAMESPACE_DIRNAME}/ruby_extension_source"
-		f.puts "nginx_module_source_dir=/usr/share/#{GLOBAL_NAMESPACE_DIRNAME}/ngx_http_passenger_module"
+		f.puts "natively_packaged=true"
+		f.puts "bin_dir=#{fake_bindir}".gsub(/#{fakeroot}/,"")
+		f.puts "agents_dir=#{fake_agents_dir}".gsub(/#{fakeroot}/,"")
+		f.puts "lib_dir=#{fake_libdir}".gsub(/#{fakeroot}/,"")
+		f.puts "helper_scripts_dir=#{fake_helper_scripts_dir}".gsub(/#{fakeroot}/,"")
+		f.puts "resources_dir=#{fake_resources_dir}".gsub(/#{fakeroot}/,"")
+		f.puts "include_dir=#{fake_include_dir}".gsub(/#{fakeroot}/,"")
+		f.puts "doc_dir=#{fake_docdir}".gsub(/#{fakeroot}/,"")
+		f.puts "ruby_libdir=#{CONFIG['sitedir']}/vendor_ruby"
+		f.puts "node_libdir=#{fake_nodelibdir}".gsub(/#{fakeroot}/,"")
+		f.puts "apache2_module_path=no"
+		f.puts "ruby_extension_source_dir=#{fake_ruby_extension_source_dir}".gsub(/#{fakeroot}/,"")
+		f.puts "nginx_module_source_dir=#{fake_nginx_module_source_dir}".gsub(/#{fakeroot}/,"")
 	end
 
 	# Sanity check the locations.ini file
