--- build/basics.rb     2013-03-05 16:55:02.000000000 +0700
+++ build/basics.rb     2013-04-03 12:42:46.012231095 +0700
@@ -134,6 +134,7 @@
 EXTRA_CXXFLAGS << " -Wno-ambiguous-member-template" if PlatformInfo.cxx_is_clang?
 EXTRA_CXXFLAGS << " #{OPTIMIZATION_FLAGS}" if !OPTIMIZATION_FLAGS.empty?
 EXTRA_CXXFLAGS << " " << string_option('EXTRA_CXXFLAGS').gsub("\n", " ") if string_option('EXTRA_CXXFLAGS')
+EXTRA_CXXFLAGS << " #{ENV['CXXFLAGS']}"

 # Extra linker flags that should always be passed to the linker.
 # These should be included first in the command string, before anything else.
--- build/packaging.rb	2013-07-09 12:10:02.000000000 +0400
+++ build/packaging.rb	2013-07-10 23:59:46.802236526 +0400
@@ -180,32 +180,27 @@
 end
 
 desc "Create a fakeroot, useful for building native packages"
-task :fakeroot => [:apache2, :nginx, :doc] do
+task :fakeroot => [:nginx, :doc] do
 	require 'rbconfig'
 	require 'fileutils'
 	include RbConfig
-	fakeroot = "pkg/fakeroot"
+	fakeroot = ENV['D']
 	
 	# We don't use CONFIG['archdir'] and the like because we want
 	# the files to be installed to /usr, and the Ruby interpreter
 	# on the packaging machine might be in /usr/local.
-	fake_rubylibdir = "#{fakeroot}/usr/lib/ruby/vendor_ruby"
-	fake_libdir = "#{fakeroot}/usr/lib/passenger"
-	fake_native_support_dir = "#{fakeroot}/usr/lib/ruby/#{CONFIG['ruby_version']}/#{CONFIG['arch']}"
-	fake_agents_dir = "#{fakeroot}/usr/lib/#{GLOBAL_NAMESPACE_DIRNAME}/agents"
-	fake_helper_scripts_dir = "#{fakeroot}/usr/share/#{GLOBAL_NAMESPACE_DIRNAME}/helper-scripts"
-	fake_resources_dir = "#{fakeroot}/usr/share/passenger"
-	fake_include_dir = "#{fakeroot}/usr/share/passenger/include"
-	fake_docdir = "#{fakeroot}/usr/share/doc/#{GLOBAL_NAMESPACE_DIRNAME}"
-	fake_bindir = "#{fakeroot}/usr/bin"
-	fake_sbindir = "#{fakeroot}/usr/sbin"
-	fake_apache2_module_dir = "#{fakeroot}/usr/lib/apache2/modules"
-	fake_apache2_module = "#{fake_apache2_module_dir}/mod_passenger.so"
-	fake_ruby_extension_source_dir = "#{fakeroot}/usr/share/passenger/ruby_extension_source"
-	
-	sh "rm -rf #{fakeroot}"
-	sh "mkdir -p #{fakeroot}"
-	
+	fake_rubylibdir = "#{fakeroot}/#{CONFIG['sitedir']}/#{CONFIG['ruby_version']}/#{CONFIG['arch']}"
+	fake_libdir = "#{fakeroot}/" + CONFIG['sitedir'].gsub(/(\/.*?\/.*?)\/.*/,"\\1") + "/#{GLOBAL_NAMESPACE_DIRNAME}"
+	fake_native_support_dir = "#{fakeroot}/#{CONFIG['sitedir']}/#{CONFIG['ruby_version']}/#{CONFIG['arch']}"
+	fake_agents_dir = "#{fakeroot}/usr/libexec/#{GLOBAL_NAMESPACE_DIRNAME}/agents"
+	fake_helper_scripts_dir = "#{fakeroot}/usr/libexec/#{GLOBAL_NAMESPACE_DIRNAME}/helper-scripts"
+	fake_resources_dir = "#{fakeroot}/usr/libexec/#{GLOBAL_NAMESPACE_DIRNAME}/resources"
+	fake_include_dir = "#{fakeroot}/usr/include/#{GLOBAL_NAMESPACE_DIRNAME}"
+	fake_docdir = "#{fakeroot}/usr/share/doc/"+ENV['P']+"/#{GLOBAL_NAMESPACE_DIRNAME}"
+	fake_bindir = "#{fakeroot}/usr/bin"
+	fake_sbindir = "#{fakeroot}/usr/sbin"
+	fake_ruby_extension_source_dir = "#{fakeroot}/#{CONFIG['sitedir']}/vendor_ruby/#{GLOBAL_NAMESPACE_DIRNAME}/ruby_extension_source"
+
 	# Ruby sources
 	sh "mkdir -p #{fake_rubylibdir}"
 	sh "cp #{PhusionPassenger.ruby_libdir}/phusion_passenger.rb #{fake_rubylibdir}/"
@@ -276,10 +271,6 @@
 		sh "cp bin/#{exe} #{fake_sbindir}/"
 	end
 	
-	# Apache 2 module
-	sh "mkdir -p #{fake_apache2_module_dir}"
-	sh "cp #{APACHE2_MODULE} #{fake_apache2_module_dir}/"
-
 	# Ruby extension sources
 	sh "mkdir -p #{fake_ruby_extension_source_dir}"
 	sh "cp -R #{PhusionPassenger.ruby_extension_source_dir}/* #{fake_ruby_extension_source_dir}"
@@ -287,17 +278,15 @@
 	puts "Creating #{fake_rubylibdir}/phusion_passenger/locations.ini"
 	File.open("#{fake_rubylibdir}/phusion_passenger/locations.ini", "w") do |f|
 		f.puts "[locations]"
-		f.puts "natively_packaged=true"
-		f.puts "bin=/usr/bin"
-		f.puts "agents=/usr/lib/passenger/agents"
-		f.puts "libdir=/usr/lib/passenger"
-		f.puts "helper_scripts=/usr/share/passenger/helper-scripts"
-		f.puts "resources=/usr/share/passenger"
-		f.puts "includedir=/usr/share/passenger/include"
-		f.puts "doc=/usr/share/doc/passenger"
-		f.puts "rubylibdir=/usr/lib/ruby/vendor_ruby"
-		f.puts "apache2_module=/usr/lib/apache2/modules/mod_passenger.so"
-		f.puts "ruby_extension_source=/usr/share/passenger/ruby_extension_source"
+		f.puts "natively_packaged=true"
+		f.puts "bin=#{fake_bindir}".gsub(/#{fakeroot}/,"")
+		f.puts "agents=#{fake_agents_dir}".gsub(/#{fakeroot}/,"")
+		f.puts "libdir=#{fake_libdir}".gsub(/#{fakeroot}/,"")
+		f.puts "helper_scripts=#{fake_helper_scripts_dir}".gsub(/#{fakeroot}/,"")
+		f.puts "resources=#{fake_resources_dir}".gsub(/#{fakeroot}/,"")
+		f.puts "includedir=#{fake_include_dir}".gsub(/#{fakeroot}/,"")
+		f.puts "doc=#{fake_docdir}".gsub(/#{fakeroot}/,"")
+		f.puts "rubylibdir=#{CONFIG['sitedir']}/vendor_ruby"
 	end
 
 	sh "find #{fakeroot} -name .DS_Store -print0 | xargs -0 rm -f"
