--- build/basics.rb	2013-04-03 12:57:45.121211033 +0700
+++ build/basics.rb	2013-04-03 13:01:29.983206016 +0700
@@ -175,6 +175,7 @@
 EXTRA_PRE_LDFLAGS  = string_option('EXTRA_PRE_LDFLAGS', '').gsub("\n", " ")
 # These should be included last in the command string, even after PlatformInfo.portability_ldflags.
 EXTRA_LDFLAGS     = string_option('EXTRA_LDFLAGS', '').gsub("\n", " ")
+EXTRA_LDFLAGS << "#{ENV['LDFLAGS']}"
 EXTRA_C_LDFLAGS   = string_option('EXTRA_C_LDFLAGS', '').gsub("\n", " ")
 EXTRA_CXX_LDFLAGS = string_option('EXTRA_CXX_LDFLAGS', '').gsub("\n", " ")
 
