--- a/Telegram/lib_ui/ui/platform/linux/ui_linux_wayland_integration_dummy.cpp
+++ b/Telegram/lib_ui/ui/platform/linux/ui_linux_wayland_integration_dummy.cpp
@@ -20,6 +20,8 @@
 WaylandIntegration::WaylandIntegration() {
 }
 
+WaylandIntegration::~WaylandIntegration() = default;
+
 WaylandIntegration *WaylandIntegration::Instance() {
 	if (!::Platform::IsWayland()) return nullptr;
 	static WaylandIntegration instance;
