--- a/Telegram/lib_webview/webview/webview_embed.cpp
+++ b/Telegram/lib_webview/webview/webview_embed.cpp
@@ -14,7 +14,6 @@
 
 #include <QtWidgets/QWidget>
 #include <QtGui/QWindow>
-#include <QtCore/QJsonDocument>
 
 namespace Webview {
 namespace {
--- a/Telegram/lib_webview/webview/webview_embed.h
+++ b/Telegram/lib_webview/webview/webview_embed.h
@@ -9,6 +9,8 @@
 #include "base/unique_qptr.h"
 #include "base/basic_types.h"
 
+#include <QtCore/QJsonDocument>
+
 class QString;
 class QWidget;
 class QWindow;
