diff -NaurBwdp a/Telegram/CMakeLists.txt b/Telegram/CMakeLists.txt
--- a/Telegram/CMakeLists.txt	2017-08-08 16:29:33.935335422 +0700
+++ b/Telegram/CMakeLists.txt	2017-08-08 16:30:35.950021360 +0700
@@ -30,8 +30,7 @@ pkg_check_modules(LIBDRM REQUIRED libdrm
 pkg_check_modules(LIBVA REQUIRED libva-x11 libva-drm libva)
 pkg_check_modules(MINIZIP REQUIRED minizip)
 
-pkg_check_modules(GTK3 REQUIRED gtk+-3.0)
-pkg_check_modules(APPINDICATOR REQUIRED appindicator3-0.1)
+add_definitions(-DTDESKTOP_DISABLE_GTK_INTEGRATION)
 
 find_package(ZLIB REQUIRED)
 
@@ -119,11 +118,9 @@ target_link_libraries(Telegram
     ${LIBVA_LIBRARIES}
     ${LIBDRM_LIBRARIES}
     ${ZLIB_LIBRARY_RELEASE}
-    ${GTK3_LIBRARIES}
     ${TGVOIP_LIBS}
     ${MINIZIP_LIBRARIES}
     ${OPENAL_LIBRARY}
-    ${APPINDICATOR_LIBRARIES}
     ${LIBLZMA_LIBRARIES}
     ${X11_X11_LIB})
 target_include_directories(Telegram PUBLIC
@@ -133,14 +130,12 @@ target_include_directories(Telegram PUBL
     ${LIBVA_INCLUDE_DIRS}
     ${LIBDRM_INCLUDE_DIRS}
     ${ZLIB_INCLUDE_DIR}
-    ${GTK3_INCLUDE_DIRS}
     ${TGVOIP_INCLUDE_DIR}
     ${GSL_INCLUDE_DIR}
     ${EMOJI_SUGGESTIONS_INCLUDE_DIR}
     ${VARIANT_INCLUDE_DIR}
     ${MINIZIP_INCLUDE_DIRS}
     ${OPENAL_INCLUDE_DIR}
-    ${APPINDICATOR_INCLUDE_DIRS}
     ${LIBLZMA_INCLUDE_DIRS}
     ${GENERATED_DIR})
 

