--- a/Telegram/lib_ui/ui/widgets/buttons.h	2025-04-02 11:00:20.160686563 +0700
+++ b/Telegram/lib_ui/ui/widgets/buttons.h	2025-04-02 11:01:57.460681639 +0700
@@ -273,7 +273,7 @@ public:
 		const style::SettingsButton &st = st::defaultSettingsButton);
 	SettingsButton(
 		QWidget *parent,
-		nullptr_t,
+		std::nullptr_t,
 		const style::SettingsButton &st = st::defaultSettingsButton);
 	~SettingsButton();
 
--- a/Telegram/lib_ui/ui/widgets/buttons.cpp	2025-04-02 11:00:25.273844239 +0700
+++ b/Telegram/lib_ui/ui/widgets/buttons.cpp	2025-04-02 11:01:57.460681639 +0700
@@ -795,7 +795,7 @@ SettingsButton::SettingsButton(
 
 SettingsButton::SettingsButton(
 	QWidget *parent,
-	nullptr_t,
+	std::nullptr_t,
 	const style::SettingsButton &st)
 : RippleButton(parent, st.ripple)
 , _st(st)
