diff -NaurBwdp a/CMakeLists.txt b/CMakeLists.txt
--- a/CMakeLists.txt	2018-05-03 02:56:40.023890420 +0700
+++ b/CMakeLists.txt	2018-05-03 03:02:29.728067372 +0700
@@ -521,71 +521,6 @@ if(UNIX)
     set(ADDITIONAL_TEST_LIBS m)
 endif(UNIX)
 
-set(SIMPLETESTS
-   cargstest
-   copyrecttest
-)
-
-if(CMAKE_USE_PTHREADS_INIT)
-  set(SIMPLETESTS
-      ${SIMPLETESTS}
-      encodingstest
-     )
-endif(CMAKE_USE_PTHREADS_INIT)
-
-file(MAKE_DIRECTORY ${CMAKE_CURRENT_BINARY_DIR}/test)
-foreach(t ${SIMPLETESTS})
-  add_executable(test_${t} ${TESTS_DIR}/${t}.c)
-  set_target_properties(test_${t} PROPERTIES OUTPUT_NAME ${t})
-  set_target_properties(test_${t} PROPERTIES RUNTIME_OUTPUT_DIRECTORY ${CMAKE_CURRENT_BINARY_DIR}/test)
-  target_link_libraries(test_${t} vncserver vncclient ${ADDITIONAL_TEST_LIBS})
-endforeach(t ${SIMPLETESTS})
-
-if(WITH_JPEG AND FOUND_LIBJPEG_TURBO)
-  add_executable(test_tjunittest
-                 ${TESTS_DIR}/tjunittest.c
-                 ${TESTS_DIR}/tjutil.c
-                 ${TESTS_DIR}/tjutil.h
-                 ${COMMON_DIR}/turbojpeg.c
-                 ${COMMON_DIR}/turbojpeg.h
-                )
-  set_target_properties(test_tjunittest PROPERTIES OUTPUT_NAME tjunittest)
-  set_target_properties(test_tjunittest PROPERTIES RUNTIME_OUTPUT_DIRECTORY ${CMAKE_CURRENT_BINARY_DIR}/test)
-  target_link_libraries(test_tjunittest vncserver vncclient ${ADDITIONAL_TEST_LIBS})
-
-  add_executable(test_tjbench
-                 ${TESTS_DIR}/tjbench.c
-                 ${TESTS_DIR}/tjutil.c
-                 ${TESTS_DIR}/tjutil.h
-                 ${TESTS_DIR}/bmp.c
-                 ${TESTS_DIR}/bmp.h
-                 ${COMMON_DIR}/turbojpeg.c
-                 ${COMMON_DIR}/turbojpeg.h
-                )
-  set_target_properties(test_tjbench PROPERTIES OUTPUT_NAME tjbench)
-  set_target_properties(test_tjbench PROPERTIES RUNTIME_OUTPUT_DIRECTORY ${CMAKE_CURRENT_BINARY_DIR}/test)
-  target_link_libraries(test_tjbench vncserver vncclient ${ADDITIONAL_TEST_LIBS})
-
-endif(WITH_JPEG AND FOUND_LIBJPEG_TURBO)
-
-if(LIBVNCSERVER_WITH_WEBSOCKETS)
-  add_executable(test_wstest
-    ${TESTS_DIR}/wstest.c
-    ${TESTS_DIR}/wstestdata.inc
-    )
-  set_target_properties(test_wstest PROPERTIES OUTPUT_NAME wstest)
-  set_target_properties(test_wstest PROPERTIES RUNTIME_OUTPUT_DIRECTORY ${CMAKE_CURRENT_BINARY_DIR}/test)
-  target_link_libraries(test_wstest vncserver vncclient ${ADDITIONAL_TEST_LIBS})
-endif(LIBVNCSERVER_WITH_WEBSOCKETS)
-
-add_test(NAME cargs COMMAND test_cargstest)
-if(FOUND_LIBJPEG_TURBO)
-    add_test(NAME turbojpeg COMMAND test_tjunittest)
-endif(FOUND_LIBJPEG_TURBO)
-if(LIBVNCSERVER_WITH_WEBSOCKETS)
-    add_test(NAME wstest COMMAND test_wstest)
-endif(LIBVNCSERVER_WITH_WEBSOCKETS)
-
 #
 # this gets the libraries needed by TARGET in "-libx -liby ..." form
 #
