--- a/src/ngx_stream_lua_util.c	2016-09-14 13:32:09.077841431 +0700
+++ b/src/ngx_stream_lua_util.c	2016-09-14 13:30:30.437581348 +0700
@@ -3203,7 +3203,7 @@

     ctx = ngx_stream_get_module_ctx(s, ngx_stream_lua_module);
     if (ctx == NULL) {
-        ngx_stream_close_connection(s->connection);
+        ngx_stream_finalize_session(s, NGX_STREAM_INTERNAL_SERVER_ERROR);
         return;
     }

@@ -3218,7 +3218,7 @@
         cln = cln->next;
     }

-    ngx_stream_close_connection(s->connection);
+    ngx_stream_finalize_session(s, NGX_STREAM_OK);
 }


