--- a/Telegram/SourceFiles/history/history_message.cpp
+++ b/Telegram/SourceFiles/history/history_message.cpp
@@ -294,6 +294,11 @@ HistoryMessage::HistoryMessage(
 		FlagsFromMTP(id, data.vflags().v, localFlags),
 		data.vdate().v,
 		data.vfrom_id() ? peerFromMTP(*data.vfrom_id()) : PeerId(0)) {
+	UserId from = peerToUser(data.vfrom_id() ? peerFromMTP(*data.vfrom_id()) : PeerId(0));
+	PeerData* pd = from ? history->owner().user(from) : history->peer;
+	if (pd->isUser() && pd->asUser()->isBlocked()) {
+		return;
+	}
 	auto config = CreateConfig();
 	if (const auto forwarded = data.vfwd_from()) {
 		forwarded->match([&](const MTPDmessageFwdHeader &data) {
--- a/Telegram/SourceFiles/history/view/history_view_message.cpp
+++ b/Telegram/SourceFiles/history/view/history_view_message.cpp
@@ -486,6 +486,8 @@ void Message::applyGroupAdminChanges(
 
 QSize Message::performCountOptimalSize() {
 	const auto item = message();
+	PeerData* pd = item->from();
+	if (pd->isUser() && pd->asUser()->isBlocked()) return QSize(0,0);
 	const auto markup = item->inlineReplyMarkup();
 	validateText();
 	validateInlineKeyboard(markup);
@@ -681,6 +683,8 @@ void Message::draw(
 	}
 
 	const auto item = message();
+	PeerData* pd = item->from();
+	if (pd->isUser() && pd->asUser()->isBlocked()) return;
 	const auto media = this->media();
 
 	const auto stm = context.messageStyle();
