--- a/config 2010-06-24 22:34:36.000000000 +0700
+++ b/config 2017-10-09 10:00:05.652603791 +0700
@@ -1,7 +1,18 @@
 ngx_addon_name=ngx_supervisord_module
 
+NGX_SRCS="$NGX_ADDON_SRCS $ngx_addon_dir/ngx_supervisord.c"
+
+if test -n "$ngx_module_link"; then
+    ngx_module_type=HTTP
+    ngx_module_name=$ngx_addon_name
+    ngx_module_incs="$ngx_addon_dir"
+    ngx_module_srcs="$NGX_SRCS"
+
+    . auto/module
+else
+    HTTP_MODULES="$HTTP_MODULES $ngx_addon_name"
 HTTP_INCS="$HTTP_INCS $ngx_addon_dir"
-HTTP_MODULES="$HTTP_MODULES ngx_supervisord_module"
-NGX_ADDON_SRCS="$NGX_ADDON_SRCS $ngx_addon_dir/ngx_supervisord.c"
+    NGX_ADDON_SRCS="$NGX_ADDON_SRCS $NGX_SRCS"
+fi
 
 have=NGX_SUPERVISORD_MODULE . auto/have
