--- a/Telegram/SourceFiles/stdafx.h
+++ b/Telegram/SourceFiles/stdafx.h
@@ -60,6 +60,8 @@
 #include <QtCore/QString>
 #include <QtCore/QStringList>
 #include <QtCore/QVector>
+#include <QtNetwork/QNetworkProxy>
+#include <QtNetwork/QTcpSocket>
 
 #include <QtGui/QIcon>
 #include <QtGui/QImage>
@@ -119,6 +121,7 @@
 
 #include "base/basic_types.h"
 #include "logs.h"
+#include "scheme.h"
 #include "core/utils.h"
 #include "config.h"
 