diff --git i/external/webrtc/CMakeLists.txt w/external/webrtc/CMakeLists.txt
index a354057..7f13b25 100644
--- i/cmake/external/webrtc/CMakeLists.txt
+++ w/cmake/external/webrtc/CMakeLists.txt
@@ -8,8 +8,10 @@ add_library(external_webrtc INTERFACE IMPORTED GLOBAL)
 add_library(desktop-app::external_webrtc ALIAS external_webrtc)

 if (DESKTOP_APP_USE_PACKAGED)
-    find_package(tg_owt REQUIRED)
-    target_link_libraries(external_webrtc INTERFACE tg_owt::tg_owt)
+    find_package(PkgConfig REQUIRED)
+    pkg_check_modules(OWT REQUIRED tg_owt)
+    target_include_directories(external_webrtc SYSTEM INTERFACE ${OWT_INCLUDE_DIRS})
+    target_link_libraries(external_webrtc INTERFACE ${OWT_LIBRARIES})
 else()
     set(webrtc_loc ${libs_loc}/tg_owt/src)
     set(webrtc_build_loc ${libs_loc}/tg_owt/out/$<CONFIG>)
