--- a/cmake/external/CMakeLists.txt
+++ a/cmake/external/CMakeLists.txt
@@ -70,3 +70,4 @@
 endif()
 add_checked_subdirectory(xxhash)
 add_checked_subdirectory(zlib)
+add_checked_subdirectory(libyuv)
--- a/cmake/external/libyuv/CMakeLists.txt
+++ b/cmake/external/libyuv/CMakeLists.txt
@@ -0,0 +1,7 @@
+if (DESKTOP_APP_USE_PACKAGED)
+    add_library(external_libyuv INTERFACE IMPORTED GLOBAL)
+    add_library(desktop-app::external_libyuv ALIAS external_libyuv)
+    find_package(PkgConfig REQUIRED)
+    pkg_check_modules(LIBYUV IMPORTED_TARGET libyuv)
+    target_link_libraries(external_libyuv INTERFACE PkgConfig::LIBYUV)
+endif()
