diff -NaurBwd a/Telegram/lib_base/base/platform/linux/base_last_input_linux.cpp b/Telegram/lib_base/base/platform/linux/base_last_input_linux.cpp
--- a/Telegram/lib_base/base/platform/linux/base_last_input_linux.cpp	2022-12-29 03:22:18.737445637 +0700
+++ b/Telegram/lib_base/base/platform/linux/base_last_input_linux.cpp	2022-12-29 03:11:09.737385713 +0700
@@ -103,7 +103,7 @@
 			"org.freedesktop.DBus.Error.AccessDenied",
 		};
 
-		const auto errorName = Gio::DBus::ErrorUtils::get_remote_error(e);
+		const auto errorName = Gio::DBus::ErrorUtils::get_remote_error(e).raw();
 		if (ranges::contains(NotSupportedErrors, errorName)) {
 			NotSupported = true;
 			return std::nullopt;
@@ -163,7 +163,7 @@
 			"org.freedesktop.DBus.Error.AccessDenied",
 		};
 
-		const auto errorName = Gio::DBus::ErrorUtils::get_remote_error(e);
+		const auto errorName = Gio::DBus::ErrorUtils::get_remote_error(e).raw();
 		if (ranges::contains(NotSupportedErrors, errorName)) {
 			NotSupported = true;
 			return std::nullopt;
diff -NaurBwd a/Telegram/SourceFiles/platform/linux/notifications_manager_linux.cpp b/Telegram/SourceFiles/platform/linux/notifications_manager_linux.cpp
--- a/Telegram/SourceFiles/platform/linux/notifications_manager_linux.cpp	2022-12-29 03:22:31.844977199 +0700
+++ b/Telegram/SourceFiles/platform/linux/notifications_manager_linux.cpp	2022-12-29 03:12:39.218178489 +0700
@@ -77,14 +77,7 @@
 			Gio::DBus::BusType::SESSION);
 
 		const auto activatable = [&] {
-			try {
-				return ranges::contains(
-					base::Platform::DBus::ListActivatableNames(connection),
-					Glib::ustring(std::string(kService)));
-			} catch (...) {
-				// avoid service restart loop in sandboxed environments
 				return true;
-			}
 		}();
 
 		return std::make_unique<base::Platform::DBus::ServiceWatcher>(
@@ -129,7 +122,7 @@
 						};
 
 						const auto errorName =
-							Gio::DBus::ErrorUtils::get_remote_error(e);
+							Gio::DBus::ErrorUtils::get_remote_error(e).raw();
 
 						if (!ranges::contains(NotSupportedErrors, errorName)) {
 							throw e;
@@ -163,13 +156,7 @@
 		}();
 
 		static const auto activatable = [&] {
-			try {
-				return ranges::contains(
-					DBus::ListActivatableNames(connection),
-					Glib::ustring(std::string(kService)));
-			} catch (...) {
-				return false;
+				return true;
-			}
 		}();
 
 		return hasOwner || activatable;
diff -NaurBwd a/Telegram/SourceFiles/platform/linux/specific_linux.cpp b/Telegram/SourceFiles/platform/linux/specific_linux.cpp
--- a/Telegram/SourceFiles/platform/linux/specific_linux.cpp	2022-12-29 03:22:37.072790370 +0700
+++ b/Telegram/SourceFiles/platform/linux/specific_linux.cpp	2022-12-29 03:12:44.891975201 +0700
@@ -262,9 +262,7 @@
 		} catch (...) {
 		}
 
-		if (ranges::contains(
-			activatableNames,
-			"org.freedesktop.Notifications")) {
+		if (true) {
 			return true;
 		}
 
