diff --git a/Telegram/SourceFiles/history/view/media/history_view_media_grouped.cpp b/Telegram/SourceFiles/history/view/media/history_view_media_grouped.cpp
index cccc8ea96..7e2293de3 100644
--- a/Telegram/SourceFiles/history/view/media/history_view_media_grouped.cpp
+++ b/Telegram/SourceFiles/history/view/media/history_view_media_grouped.cpp
@@ -19,6 +19,7 @@
 #include "ui/grouped_layout.h"
 #include "ui/text_options.h"
 #include "layout.h"
+#include "facades.h"
 #include "styles/style_history.h"
 
 namespace HistoryView {
@@ -76,9 +77,17 @@ QSize GroupedMedia::countOptimalSize() {
 		sizes.push_back(media->sizeForGrouping());
 	}
 
+	const auto captionWithPaddings = _caption.maxWidth()
+			+ st::msgPadding.left()
+			+ st::msgPadding.right();
+	auto groupMaxWidth = st::historyGroupWidthMax;
+	if (Adaptive::ChatWide()) {
+		accumulate_max(groupMaxWidth, captionWithPaddings);
+	}
+
 	const auto layout = Ui::LayoutMediaGroup(
 		sizes,
-		st::historyGroupWidthMax,
+		groupMaxWidth,
 		st::historyGroupWidthMin,
 		st::historyGroupSkip);
 	Assert(layout.size() == _parts.size());
@@ -94,6 +103,9 @@ QSize GroupedMedia::countOptimalSize() {
 	}
 
 	if (!_caption.isEmpty()) {
+		if (Adaptive::ChatWide()) {
+			maxWidth = qMax(maxWidth, captionWithPaddings);
+		}
 		auto captionw = maxWidth - st::msgPadding.left() - st::msgPadding.right();
 		minHeight += st::mediaCaptionSkip + _caption.countHeight(captionw);
 		if (isBubbleBottom()) {
