diff --git a/Telegram/SourceFiles/history/view/media/history_view_web_page.cpp b/Telegram/SourceFiles/history/view/media/history_view_web_page.cpp
index 6a096c283..eef78067f 100644
--- a/Telegram/SourceFiles/history/view/media/history_view_web_page.cpp
+++ b/Telegram/SourceFiles/history/view/media/history_view_web_page.cpp
@@ -24,6 +24,7 @@
 #include "data/data_photo.h"
 #include "data/data_file_origin.h"
 #include "app.h"
+#include "facades.h"
 #include "styles/style_history.h"
 
 namespace HistoryView {
@@ -278,6 +279,10 @@ QSize WebPage::countOptimalSize() {
 		_durationWidth = st::msgDateFont->width(_duration);
 	}
 	maxWidth += st::msgPadding.left() + st::webPageLeft + st::msgPadding.right();
+	if (Adaptive::ChatWide()) {
+		accumulate_min(maxWidth, st::msgMaxWidth);
+		accumulate_max(maxWidth, _parent->plainMaxWidth());
+	}
 	auto padding = inBubblePadding();
 	minHeight += padding.top() + padding.bottom();
 
@@ -292,6 +297,10 @@ QSize WebPage::countCurrentSize(int newWidth) {
 		return { newWidth, minHeight() };
 	}
 
+	if (Adaptive::ChatWide() && !asArticle()) {
+		accumulate_min(newWidth, maxWidth());
+	}
+
 	auto innerWidth = newWidth - st::msgPadding.left() - st::webPageLeft - st::msgPadding.right();
 	auto newHeight = 0;
 
