diff -NaurBwdp a/config b/config
--- a/config	2017-10-12 04:55:03.223753991 +0700
+++ b/config	2017-10-12 05:13:44.218216528 +0700
@@ -1,15 +1,25 @@
 # Copyright (C) Valentin V. Bartenev
 ngx_addon_name=ngx_http_ctpp2_module
 
+ngx_ctpp_srcs="$ngx_addon_dir/sources/CTPP2NginxVMEnvironment.cpp $ngx_addon_dir/sources/ctpp2_process.cpp $ngx_addon_dir/sources/ngx_http_ctpp2_filter_module.c $ngx_addon_dir/sources/ngx_http_ctpp2_tmpl_loader.c"
+ngx_ctpp_libs="-lstdc++ -lctpp2"
+
+if test -n "$ngx_module_link"; then
+    ngx_module_type=HTTP_FILTER
+    ngx_module_name=$ngx_addon_name
+    ngx_module_srcs="$ngx_ctpp_srcs"
+    ngx_module_libs="$ngx_ctpp_libs"
+    ngx_module_order="ngx_http_ctpp2_filter_module ngx_http_copy_filter_module ngx_http_ctpp2_tmpl_loader"
+
+    . auto/module
+else
 HTTP_COPY_FILTER_MODULE="ngx_http_ctpp2_filter_module $HTTP_COPY_FILTER_MODULE ngx_http_ctpp2_tmpl_loader"
 
 NGX_ADDON_SRCS="$NGX_ADDON_SRCS
-	$ngx_addon_dir/sources/CTPP2NginxVMEnvironment.cpp
-	$ngx_addon_dir/sources/ctpp2_process.cpp
-	$ngx_addon_dir/sources/ngx_http_ctpp2_filter_module.c
-	$ngx_addon_dir/sources/ngx_http_ctpp2_tmpl_loader.c"
+	$ngx_ctpp_srcs"
 
-CORE_LIBS="$CORE_LIBS -lstdc++ -lctpp2"
+CORE_LIBS="$CORE_LIBS $ngx_ctpp_libs"
+fi
 
 TMPLS_ROOT_PATH=${TMPLS_ROOT_PATH:-templates}
 have=NGX_CTPP2_TMPLS_ROOT_PATH value="\"$TMPLS_ROOT_PATH\"" . auto/define

