diff -NaurBwdp a/config b/config
--- a/config	2017-10-12 01:20:47.479817195 +0700
+++ b/config	2017-10-12 01:25:45.751330529 +0700
@@ -9,9 +9,17 @@ ngx_feature_test="rrd_graph_v;"
 
 if [ $ngx_found = yes ]; then
     ngx_addon_name=ngx_http_rrd_graph_module
+    if test -n "$ngx_module_link"; then
+        ngx_module_type=HTTP
+        ngx_module_name=$ngx_addon_name
+        ngx_module_srcs=$ngx_addon_dir/ngx_http_rrd_graph_module.c
+        ngx_module_libs="-lrrd"
+        . auto/module
+    else
     HTTP_MODULES="$HTTP_MODULES ngx_http_rrd_graph_module"
     NGX_ADDON_SRCS="$NGX_ADDON_SRCS $ngx_addon_dir/ngx_http_rrd_graph_module.c"
     CORE_LIBS="$CORE_LIBS -lrrd"
+    fi
 else
     cat << END
 $0: error: ngx_rrd_graph requires RRDtool 1.3 or later.
