diff -NaurBwd a/Telegram/SourceFiles/history/history_item.cpp b/Telegram/SourceFiles/history/history_item.cpp
--- a/Telegram/SourceFiles/history/history_item.cpp	2017-05-23 01:26:09.007253615 +0700
+++ b/Telegram/SourceFiles/history/history_item.cpp	2017-05-23 01:26:16.305983785 +0700
@@ -519,13 +519,6 @@
 	p.setFont(st::historyUnreadBarFont);
 	p.setPen(st::historyUnreadBarFg);
 
-	int left = st::msgServiceMargin.left();
-	int maxwidth = w;
-	if (Adaptive::ChatWide()) {
-		maxwidth = qMin(maxwidth, int32(st::msgMaxWidth + 2 * st::msgPhotoSkip + 2 * st::msgMargin.left()));
-	}
-	w = maxwidth;
-
 	p.drawText((w - _width) / 2, y + marginTop() + (st::historyUnreadBarHeight - 2 * st::lineWidth - st::historyUnreadBarFont->height) / 2 + st::historyUnreadBarFont->ascent, _text);
 }
 
diff -NaurBwd a/Telegram/SourceFiles/history/history_media_types.cpp b/Telegram/SourceFiles/history/history_media_types.cpp
--- a/Telegram/SourceFiles/history/history_media_types.cpp	2017-05-23 01:26:09.007253615 +0700
+++ b/Telegram/SourceFiles/history/history_media_types.cpp	2017-05-23 01:26:16.305983785 +0700
@@ -262,7 +262,7 @@
 	if (!tw || !th) {
 		tw = th = 1;
 	}
-	if (tw > st::maxMediaSize) {
+	if (!Adaptive::ChatWide() && tw > st::maxMediaSize) {
 		th = (st::maxMediaSize * th) / tw;
 		tw = st::maxMediaSize;
 	}
@@ -299,7 +299,7 @@
 	bool bubble = _parent->hasBubble();
 
 	int tw = convertScale(_data->full->width()), th = convertScale(_data->full->height());
-	if (tw > st::maxMediaSize) {
+	if (!Adaptive::ChatWide() && tw > st::maxMediaSize) {
 		th = (st::maxMediaSize * th) / tw;
 		tw = st::maxMediaSize;
 	}
@@ -1053,8 +1053,12 @@
 
 	if (auto named = Get<HistoryDocumentNamed>()) {
 		_maxw = qMax(tleft + named->_namew + tright, _maxw);
+		if (Adaptive::ChatWide() && captioned) {
+			_maxw = qMax(captioned->_caption.maxWidth(), _maxw);
+		} else {
 		_maxw = qMin(_maxw, int(st::msgMaxWidth));
 	}
+	}
 
 	if (thumbed) {
 		_minh = st::msgFileThumbPadding.top() + st::msgFileThumbSize + st::msgFileThumbPadding.bottom();
diff -NaurBwd a/Telegram/SourceFiles/history/history_message.cpp b/Telegram/SourceFiles/history/history_message.cpp
--- a/Telegram/SourceFiles/history/history_message.cpp	2017-05-23 01:26:09.007253615 +0700
+++ b/Telegram/SourceFiles/history/history_message.cpp	2017-05-23 01:26:16.305983785 +0700
@@ -921,7 +921,11 @@ bool HistoryMessage::drawBubble() const {
 }
 
 QRect HistoryMessage::countGeometry() const {
-	auto maxwidth = qMin(st::msgMaxWidth, _maxw);
+	auto maxwidth = _maxw;
+	if (!Adaptive::ChatWide()) {
+		maxwidth = qMin(int(st::msgMaxWidth), maxwidth);
+	}
+
 	if (_media && _media->currentWidth() < maxwidth) {
 		maxwidth = qMax(_media->currentWidth(), qMin(maxwidth, plainMaxWidth()));
 	}
@@ -1528,7 +1532,7 @@ int HistoryMessage::performResizeGetHeight() {
 	auto contentWidth = width() - (st::msgMargin.left() + st::msgMargin.right());
 	if (contentWidth < st::msgPadding.left() + st::msgPadding.right() + 1) {
 		contentWidth = st::msgPadding.left() + st::msgPadding.right() + 1;
-	} else if (contentWidth > st::msgMaxWidth) {
+	} else if (contentWidth > st::msgMaxWidth && !Adaptive::ChatWide()) {
 		contentWidth = st::msgMaxWidth;
 	}
 	if (drawBubble()) {
--- a/Telegram/SourceFiles/history/history_service.cpp
+++ b/Telegram/SourceFiles/history/history_service.cpp
@@ -413,7 +413,7 @@ bool HistoryService::updateDependencyIte
 QRect HistoryService::countGeometry() const {
 	auto result = QRect(0, 0, width(), _height);
 	if (Adaptive::ChatWide()) {
-		result.setWidth(qMin(result.width(), st::msgMaxWidth + 2 * st::msgPhotoSkip + 2 * st::msgMargin.left()));
+		result.setWidth(result.width() - st::msgServiceMargin.left() - st::msgServiceMargin.left());
 	}
 	return result.marginsRemoved(st::msgServiceMargin);
 }
@@ -482,9 +482,6 @@ int HistoryService::resizeContentGetHeig
 		_textHeight = 0;
 	} else {
 		auto contentWidth = width();
-		if (Adaptive::ChatWide()) {
-			accumulate_min(contentWidth, st::msgMaxWidth + 2 * st::msgPhotoSkip + 2 * st::msgMargin.left());
-		}
 		contentWidth -= st::msgServiceMargin.left() + st::msgServiceMargin.left(); // two small margins
 		if (contentWidth < st::msgServicePadding.left() + st::msgServicePadding.right() + 1) {
 			contentWidth = st::msgServicePadding.left() + st::msgServicePadding.right() + 1;
diff --git a/Telegram/SourceFiles/history/history_service_layout.cpp b/Telegram/SourceFiles/history/history_service_layout.cpp
index 4771f21ff..615056b71 100644
--- a/Telegram/SourceFiles/history/history_service_layout.cpp
+++ b/Telegram/SourceFiles/history/history_service_layout.cpp
@@ -165,11 +165,7 @@ void paintBubblePart(Painter &p, int x, int y, int width, int height, SideStyle
 
 void paintPreparedDate(Painter &p, const QString &dateText, int dateTextWidth, int y, int w) {
 	int left = st::msgServiceMargin.left();
-	int maxwidth = w;
-	if (Adaptive::ChatWide()) {
-		maxwidth = qMin(maxwidth, WideChatWidth());
-	}
-	w = maxwidth - st::msgServiceMargin.left() - st::msgServiceMargin.left();
+	w = w - st::msgServiceMargin.left() - st::msgServiceMargin.left();
 
 	left += (w - dateTextWidth - st::msgServicePadding.left() - st::msgServicePadding.right()) / 2;
 	int height = st::msgServicePadding.top() + st::msgServiceFont->height + st::msgServicePadding.bottom();
diff --git a/Telegram/SourceFiles/historywidget.cpp b/Telegram/SourceFiles/historywidget.cpp
index be1ba85f1..973e946c1 100644
--- a/Telegram/SourceFiles/history/history_widget.cpp
+++ b/Telegram/SourceFiles/history/history_widget.cpp
@@ -623,7 +623,13 @@
 
 	connect(&_updateEditTimeLeftDisplay, SIGNAL(timeout()), this, SLOT(updateField()));
 
-	subscribe(Adaptive::Changed(), [this] { update(); });
+	subscribe(Adaptive::Changed(), [this] {
+		if (_history) {
+			_history->setPendingResize();
+		}
+		updateHistoryGeometry(false, false, {ScrollChangeAdd, App::main() ? App::main()->contentScrollAddToY() : 0});
+		update();
+	});
 	subscribe(Global::RefItemRemoved(), [this](HistoryItem *item) {
 		itemRemoved(item);
 	});
diff --git a/Telegram/SourceFiles/profile/profile_info_widget.cpp b/Telegram/SourceFiles/profile/profile_info_widget.cpp
index bbb62d9e2..579243e75 100644
--- a/Telegram/SourceFiles/profile/profile_block_info.cpp
+++ b/Telegram/SourceFiles/profile/profile_block_info.cpp
@@ -72,7 +72,9 @@ int InfoWidget::resizeGetHeight(int newWidth) {
 		int availableWidth = newWidth - left - st::profileBlockMarginRight;
 		int maxWidth = st::msgMaxWidth;
 		accumulate_min(textWidth, availableWidth);
-		accumulate_min(textWidth, st::msgMaxWidth);
+		if (!Adaptive::ChatWide()) {
+			accumulate_min(textWidth, st::msgMaxWidth);
+		}
 		_about->resizeToWidth(textWidth + marginLeft + marginRight);
 		_about->moveToLeft(left - marginLeft, newHeight - st::profileBlockTextPart.margin.top());
 		newHeight += _about->height();
