diff -NaurBwd a/build/CMakeLists.txt b/build/CMakeLists.txt
--- a/build/CMakeLists.txt	2017-05-11 23:56:52.502311267 +0700
+++ b/build/CMakeLists.txt	2017-05-11 23:57:13.208507499 +0700
@@ -110,7 +110,7 @@
 )
 
 if(WITH_WEBSOCKETS)
-  list (APPEND CLIENT_SRC "${LIBI2PD_SRC_DIR}/Websocket.cpp")
+  list (APPEND CLIENT_SRC "${LIBI2PD_CLIENT_SRC_DIR}/Websocket.cpp")
 endif ()
 add_library(i2pdclient ${CLIENT_SRC})
 

@@ -123,6 +123,9 @@
   "${DAEMON_SRC_DIR}/i2pd.cpp"
   "${DAEMON_SRC_DIR}/UPnP.cpp"
 )
+if(WITH_WEBSOCKETS)
+	list (INSERT DAEMON_SRC 0 "${LIBI2PD_CLIENT_SRC_DIR}/Websocket.cpp")
+endif ()
 
 if (WITH_MESHNET)
   add_definitions(-DMESHNET)

