diff --git a/prosody.cfg.lua.dist b/prosody.cfg.lua.dist
index e513b11..08ebca6 100644
--- a/prosody.cfg.lua.dist
+++ b/prosody.cfg.lua.dist
@@ -123,6 +123,8 @@ s2s_secure_auth = false
 
 -- Select the authentication backend to use. The 'internal' providers
 -- use Prosody's configured data storage to store the authentication data.
+-- See http://prosody.im/doc/authentication for other possibilities including
+-- Cyrus SASL.
 
 authentication = "internal_hashed"
 
@@ -153,12 +155,19 @@ archive_expires_after = "1w" -- Remove a
 -- Logging configuration
 -- For advanced logging see https://prosody.im/doc/logging
 log = {
-	info = "prosody.log"; -- Change 'info' to 'debug' for verbose logging
-	error = "prosody.err";
-	-- "*syslog"; -- Uncomment this for logging to syslog
+	info = "/var/log/jabber/prosody.log"; -- Change 'info' to 'debug' for verbose logging
+	error = "/var/log/jabber/prosody.err";
+	-- "*syslog"; -- Uncomment this for logging to syslog; needs mod_posix
 	-- "*console"; -- Log to the console, useful for debugging with daemonize=false
 }
 
+-- Pidfile, used by prosodyctl and the init.d script
+pidfile = "/var/run/jabber/prosody.pid"
+
+-- User and group, used for daemon
+prosody_user = "jabber"
+prosody_group = "jabber"
+
 -- Uncomment to enable statistics
 -- For more info see https://prosody.im/doc/statistics
 -- statistics = "internal"
