--- a/CMakeLists.txt	2020-02-07 01:36:55.030420736 +0700
+++ b/CMakeLists.txt	2020-02-07 01:36:50.558574561 +0700
@@ -8,7 +8,7 @@ project(tl-expected
 include(CMakePackageConfigHelpers)
 include(CMakeDependentOption)
 include(GNUInstallDirs)
-include(FetchContent)
+#include(FetchContent)
 include(CTest)
 
 if (NOT DEFINED CMAKE_CXX_STANDARD)
@@ -67,9 +67,9 @@ if(EXPECTED_BUILD_TESTS)
   set(CATCH_INSTALL_HELPERS OFF)
   set(CATCH_BUILD_TESTING OFF)
   set(CATCH_INSTALL_DOCS OFF)
-  FetchContent_Declare(Catch2 URL
-    https://github.com/catchorg/Catch2/archive/v2.9.2.zip) 
-  FetchContent_MakeAvailable(Catch2)
+#  FetchContent_Declare(Catch2 URL
+#    https://github.com/catchorg/Catch2/archive/v2.9.2.zip) 
+#  FetchContent_MakeAvailable(Catch2)
 
   file(GLOB test-sources CONFIGURE_DEPENDS tests/*.cpp)
   list(FILTER test-sources EXCLUDE REGEX "tests/test.cpp")
