diff --git a/cmake/external/td/CMakeLists.txt b/cmake/external/td/CMakeLists.txt
index fa4555e8..15b276d0 100644
--- a/cmake/external/td/CMakeLists.txt
+++ b/cmake/external/td/CMakeLists.txt
@@ -7,6 +7,12 @@
 add_library(external_td INTERFACE IMPORTED GLOBAL)
 add_library(desktop-app::external_td ALIAS external_td)
 
+if (DESKTOP_APP_USE_PACKAGED)
+    find_package(Td REQUIRED)
+    target_link_libraries(external_td INTERFACE Td::tde2e)
+    return()
+endif()
+
 set(td_loc ${libs_loc}/td)
 set(td_build_loc ${td_loc}/out/$<IF:$<CONFIG:Debug>,Debug,Release>)
 if (WIN32)
@@ -19,7 +25,6 @@ endif()
 
 target_include_directories(external_td
 INTERFACE
-    ${td_loc}
     ${td_loc}/tde2e
     ${td_loc}/tdutils
     ${td_build_loc}/tdutils

From efc4f76b326172e9799e40843f83bec10ad199d6 Mon Sep 17 00:00:00 2001
From: Ilya Fedin <fedin-ilja2010@ya.ru>
Date: Sat, 3 May 2025 02:21:40 +0400
Subject: [PATCH 3/5] Add lz4 cmake config support

---
 external/lz4/CMakeLists.txt | 6 ++++++
 1 file changed, 6 insertions(+)

diff --git a/cmake/external/lz4/CMakeLists.txt b/cmake/external/lz4/CMakeLists.txt
index d55e0046..f1451e78 100644
--- a/cmake/external/lz4/CMakeLists.txt
+++ b/cmake/external/lz4/CMakeLists.txt
@@ -8,6 +8,12 @@ add_library(external_lz4 INTERFACE IMPORTED GLOBAL)
 add_library(desktop-app::external_lz4 ALIAS external_lz4)
 
 if (DESKTOP_APP_USE_PACKAGED)
+    find_package(lz4 QUIET)
+    if (lz4_FOUND)
+        target_link_libraries(external_lz4 INTERFACE lz4::lz4)
+        return()
+    endif()
+
     find_package(PkgConfig REQUIRED)
     if (PkgConfig_FOUND)
         pkg_check_modules(LZ4 IMPORTED_TARGET liblz4)

From c82c0ab8fc0f9fca0c39c665a658b070aa5661ff Mon Sep 17 00:00:00 2001
From: Ilya Fedin <fedin-ilja2010@ya.ru>
Date: Sat, 3 May 2025 02:21:47 +0400
Subject: [PATCH 4/5] Add Opus cmake config support

---
 external/opus/CMakeLists.txt | 6 ++++++
 1 file changed, 6 insertions(+)

diff --git a/cmake/external/opus/CMakeLists.txt b/cmake/external/opus/CMakeLists.txt
index 54da4eba..92336151 100644
--- a/cmake/external/opus/CMakeLists.txt
+++ b/cmake/external/opus/CMakeLists.txt
@@ -8,6 +8,12 @@ if (DESKTOP_APP_USE_PACKAGED)
     add_library(external_opus INTERFACE IMPORTED GLOBAL)
     add_library(desktop-app::external_opus ALIAS external_opus)
 
+    find_package(Opus QUIET)
+    if (Opus_FOUND)
+        target_link_libraries(external_opus INTERFACE Opus::opus)
+        return()
+    endif()
+
     find_package(PkgConfig REQUIRED REQUIRED)
     pkg_check_modules(OPUS IMPORTED_TARGET opus)
 

From e76dce0783ac7927f77d046f6ad757df442653cd Mon Sep 17 00:00:00 2001
From: Ilya Fedin <fedin-ilja2010@ya.ru>
Date: Sat, 3 May 2025 02:21:55 +0400
Subject: [PATCH 5/5] Add xxHash cmake config support

---
 external/xxhash/CMakeLists.txt | 6 ++++++
 1 file changed, 6 insertions(+)

diff --git a/cmake/external/xxhash/CMakeLists.txt b/cmake/external/xxhash/CMakeLists.txt
index b55d2698..323872a0 100644
--- a/cmake/external/xxhash/CMakeLists.txt
+++ b/cmake/external/xxhash/CMakeLists.txt
@@ -8,6 +8,12 @@ add_library(external_xxhash INTERFACE IMPORTED GLOBAL)
 add_library(desktop-app::external_xxhash ALIAS external_xxhash)
 
 if (DESKTOP_APP_USE_PACKAGED)
+    find_package(xxHash QUIET)
+    if (xxHash_FOUND)
+        target_link_libraries(external_xxhash INTERFACE xxHash::xxhash)
+        return()
+    endif()
+
     find_package(PkgConfig REQUIRED)
     if (PkgConfig_FOUND)
         pkg_check_modules(XXHASH IMPORTED_TARGET libxxhash)
