diff --git a/ngx_http_sticky_module.c b/ngx_http_sticky_module.c
index 9967428..f13d4b5 100644
--- a/ngx_http_sticky_module.c
+++ b/ngx_http_sticky_module.c
@@ -10,9 +10,7 @@
 
 #include "ngx_http_sticky_misc.h"
 
-#if (NGX_UPSTREAM_CHECK_MODULE)
-#include "ngx_http_upstream_check_handler.h"
-#endif
+#include "ngx_http_upstream_check_module.h"
 
 
 /* define a peer */
@@ -297,6 +297,14 @@ static ngx_int_t ngx_http_get_sticky_peer(ngx_peer_connection_t *pc, void *data)
 					return NGX_BUSY;
 				}
 
+                ngx_log_debug1(NGX_LOG_DEBUG_HTTP, pc->log, 0,
+                               "get sticky peer, check_index: %ui",
+                               peer->check_index);
+
+                if (ngx_http_upstream_check_peer_down(peer->check_index)) {
+                    return NGX_BUSY;
+                }
+
 				/* if it's been ignored for long enought (fail_timeout), reset timeout */
 				/* do this check before testing peer->fails ! :) */
 				if (now - peer->accessed > peer->fail_timeout) {
@@ -313,6 +323,12 @@ static ngx_int_t ngx_http_get_sticky_peer(ngx_peer_connection_t *pc, void *data)
 			/* ensure the peer is not marked as down */
 			if (!peer->down) {
 
+                ngx_log_debug1(NGX_LOG_DEBUG_HTTP, pc->log, 0,
+                               "get sticky peer, check_index: %ui",
+                               peer->check_index);
+
+                if (!ngx_http_upstream_check_peer_down(peer->check_index)) {
+
 				/* if it's not failedi, use it */
 				if (peer->max_fails == 0 || peer->fails < peer->max_fails) {
 					selected_peer = (ngx_int_t)n;
@@ -327,6 +345,7 @@ static ngx_int_t ngx_http_get_sticky_peer(ngx_peer_connection_t *pc, void *data)
 					/* mark the peer as tried */
 					iphp->rrp.tried[n] |= m;
 				}
+		}
 			}
 		}
 	}
