--- a/Telegram/SourceFiles/platform/linux/file_utilities_linux.cpp
+++ b/Telegram/SourceFiles/platform/linux/file_utilities_linux.cpp
@@ -194,9 +194,6 @@ bool GetNative(QStringList &files, QByteArray &remoteContent, const QString &cap
 } // namespace
 
 bool Get(QStringList &files, QByteArray &remoteContent, const QString &caption, const QString &filter, Type type, QString startFile) {
-	if (NativeSupported()) {
-		return GetNative(files, remoteContent, caption, filter, type, startFile);
-	}
 	return ::FileDialog::internal::GetDefault(files, remoteContent, caption, filter, type, startFile);
 }
 
