--- /etc/ld.so.conf.d/wwwwetoken-ld.conf.orig   2010-02-27 10:31:30.000000000 +0100
+++ /etc/ld.so.conf.d/wwwwetoken-ld.conf        2010-02-27 10:31:53.000000000 +0100
@@ -1,2 +1,2 @@
-/usr/lib/eToken/
+LDPATH="/usr/lib/eToken/:/usr/lib32/eToken/"

--- /etc/init.d/eTSrv.orig	2010-02-27 09:11:07.000000000 +0100
+++ /etc/init.d/eTSrv	2010-02-27 12:16:01.000000000 +0100
@@ -1,4 +1,7 @@
-#!/bin/bash
+#!/sbin/runscript
+
+# Distributed under the terms of the GNU General Purpose License v2
+# $Header: $
 # 
 # eTSrv	This starts, stops, and reloads the eToken eTSrv
 #
@@ -9,6 +12,7 @@
 #
 # Chanelog:
 # 2007-06-26 - First version of this file Nir Shahaf <nshahaf@swiftouch.com>
+# 2010-02-27 - Adapt to Gentoo Bartosz Biegun <bartosz.biegun@gmail.com>
 
 PATH=/sbin:/usr/sbin:$PATH
 RETVAL=0
@@ -16,63 +20,47 @@
 PID_F=/var/run/$prog.pid
 SIG="-1"
 
-
-# Check if we're root
-if [ ! `id -u` = 0 ]
-then
-	echo "Are you root?"
-	exit
-fi
-
-usage ()
-{
-    echo "Usage: service $prog {start|stop|restart}"
-    RETVAL=1
-}
-
-
 start ()
 {
     # Check if we're already running
-    echo -n $"Starting $prog: "
+    ebegin $"Starting $prog: "
     running
     if [ $RETVAL == "1" ]
     then
-	echo -n $"$prog already running "
-	echo
-	exit
+	ewarn $"$prog already running "
+	eend 0
     else
 	killing_clean
 	/usr/bin/eTSrv &
+
+	a=$?
+	[ $a != 0 ] && eend $a
+
 	pid=`pidof -o $$ -o $PPID -o %PPID -x $prog`
 	if [ -n "$pid" ]
 	then
 		echo "$pid" > /var/run/$prog.pid
-		echo -n $"Done"
-		echo
-		RETVAL=0
+		eend 0
 	else
-		RETVAL=1
-		echo -n $"Failed"
-		echo
-		exit
+		eend 1
 	fi
     fi
+    eend $RETVAL
 }
 
 stop ()
 {
     RETVAL=0
-    echo -n $"Stopping $prog: "
+    ebegin $"Stopping $prog: "
     not_running
     killing
     if [ $RETVAL == "0" ]
     then
-	echo -n $"Done"
+	    eend 0
     else
-	echo -n $"Failed to stop $prog "
+	    eend 1
     fi
-    echo
+    eend $RETVAL
 }
 
 restart ()
@@ -95,9 +83,8 @@
     pid=`pidof -o $$ -o $PPID -o %PPID -x $prog`
     if [ -z "$pid" ]
     then
-        echo -n $"$prog not running, nothing to do"
-	echo
-	exit
+        ewarn $"$prog not running, nothing to do"
+	eend 0
     fi
 }
 
@@ -110,12 +97,16 @@
 	for prog_id in `cat $PID_F`
 	do
 		kill $SIG $prog_id 2>&1 > /dev/null
+		a=$?
+		[ $a != 0 ] && eend $a
 	done
     else
 	prog_id=`pidof -o $$ -o $PPID -o %PPID -x $prog`
 	if [ ! -z $prog_id ]
 	then
 		kill $SIG $prog_id 2>&1 > /dev/null
+		a=$?
+		[ $a != 0 ] && eend $a
 	fi
     fi
 }
@@ -129,6 +120,7 @@
 	rm -f /tmp/$prog 2>&1 > /dev/null
     fi
 }
+
 killing ()
 {
     killing_check
@@ -148,11 +140,3 @@
     fi
 }
 
-case "$1" in
-    start) start ;;
-    stop) stop ;;
-    restart) restart ;;
-    *) usage ;;
-esac
-
-exit $RETVAL
