--- app/controllers/application_controller.rb
+++ app/controllers/application_controller.rb
@@ -19,6 +19,7 @@ require 'uri'
 require 'cgi'
 
 class ApplicationController < ActionController::Base
+  include SimpleCaptcha::ControllerHelpers
   include Redmine::I18n
 
   layout 'base'
--- config/routes.rb
+++ config/routes.rb
@@ -269,4 +269,6 @@ ActionController::Routing::Routes.draw do |map|
   map.connect 'robots.txt', :controller => 'welcome', :action => 'robots'
   # Used for OpenID
   map.root :controller => 'account', :action => 'login'
+  # Captcha
+  map.simple_captcha '/simple_captcha/:action', :controller => 'simple_captcha'
 end
--- vendor/plugins/simple_captcha/lib/simple_captcha_config.rb
+++ vendor/plugins/simple_captcha/lib/simple_captcha_config.rb
@@ -12,7 +12,7 @@ module SimpleCaptcha #:nodoc
     end
     
     def simple_captcha_key #:nodoc
-      session[:simple_captcha] ||= Digest::SHA1.hexdigest(Time.now.to_s + session.session_id.to_s)
+      session[:simple_captcha] ||= Digest::SHA1.hexdigest(Time.now.to_s)
     end
         
     def simple_captcha_value(key = simple_captcha_key) #:nodoc
--- app/controllers/account_controller.rb
+++ app/controllers/account_controller.rb
@@ -92,6 +92,7 @@ class AccountController < ApplicationController
       @user = User.new(params[:user])
       @user.admin = false
       @user.status = User::STATUS_REGISTERED
+      captcha_valid = simple_captcha_valid?
       if session[:auth_source_registration]
         @user.status = User::STATUS_ACTIVE
         @user.login = session[:auth_source_registration][:login]
@@ -106,13 +107,18 @@ class AccountController < ApplicationController
         @user.login = params[:user][:login]
         @user.password, @user.password_confirmation = params[:password], params[:password_confirmation]

-        case Setting.self_registration
-        when '1'
-          register_by_email_activation(@user)
-        when '3'
-          register_automatically(@user)
+        if !captcha_valid
+          @user.valid?
+          @user.errors.add :base, l(:notice_account_invalid_captcha)
         else
-          register_manually_by_administrator(@user)
+          case Setting.self_registration
+          when '1'
+            register_by_email_activation(@user)
+          when '3'
+            register_automatically(@user)
+          else
+            register_manually_by_administrator(@user)
+          end
         end
       end
     end
--- app/views/account/register.rhtml
+++ app/views/account/register.rhtml
@@ -37,6 +37,9 @@
 <% @user.custom_field_values.select {|v| v.editable? || v.required?}.each do |value| %>
        <p><%= custom_field_tag_with_label :user, value %></p>
 <% end %>
+
+<p><span><label for="captcha"><%=l(:field_captcha_valid)%> <span class="required">*</span></label>
+<%= show_simple_captcha(:label => '') %></span></p>
 <!--[eoform:user]-->
 </div>
 
