diff --git a/config b/config
index 6f27634..c1e8aee 100644
--- a/config
+++ b/config
@@ -1,7 +1,19 @@
 ngx_addon_name=ngx_coolkit_module
 
+NGX_SRCS="$ngx_addon_dir/src/ngx_coolkit_handlers.c $ngx_addon_dir/src/ngx_coolkit_module.c $ngx_addon_dir/src/ngx_coolkit_variables.c"
+NGX_DEPS="$ngx_addon_dir/src/ngx_coolkit_handlers.h $ngx_addon_dir/src/ngx_coolkit_module.h $ngx_addon_dir/src/ngx_coolkit_variables.h"
+
+if test -n "$ngx_module_link"; then
+    ngx_module_type=HTTP
+    ngx_module_name=$ngx_addon_name
+    ngx_module_srcs="$NGX_SRCS"
+    ngx_module_deps="$NGX_DEPS"
+
+    . auto/module
+else
-HTTP_MODULES="$HTTP_MODULES ngx_coolkit_module"
-NGX_ADDON_SRCS="$NGX_ADDON_SRCS $ngx_addon_dir/src/ngx_coolkit_handlers.c $ngx_addon_dir/src/ngx_coolkit_module.c $ngx_addon_dir/src/ngx_coolkit_variables.c"
-NGX_ADDON_DEPS="$NGX_ADDON_DEPS $ngx_addon_dir/src/ngx_coolkit_handlers.h $ngx_addon_dir/src/ngx_coolkit_module.h $ngx_addon_dir/src/ngx_coolkit_variables.h"
+    HTTP_MODULES="$HTTP_MODULES $ngx_coolkit_module"
+    NGX_ADDON_SRCS="$NGX_ADDON_SRCS $NGX_SRCS"
+    NGX_ADDON_DEPS="$NGX_ADDON_DEPS $NGX_DEPS"
+fi
 
 have=NGX_COOLKIT_MODULE . auto/have

