--- a/Telegram/ThirdParty/libtgvoip/CMakeLists.txt	2017-05-13 21:16:42.822937665 +0700
+++ b/Telegram/ThirdParty/libtgvoip/CMakeLists.txt	2017-05-14 05:27:22.694735012 +0700
@@ -8,7 +8,6 @@
 
 find_package(PkgConfig REQUIRED)
 pkg_check_modules(OPUS REQUIRED opus)
-pkg_check_modules(LIBPULSE REQUIRED libpulse)
 
 
 set(TARGET "libtgvoip")
@@ -76,9 +75,6 @@
   "NetworkSocket.cpp"
   "os/linux/AudioInputALSA.cpp"
   "os/linux/AudioOutputALSA.cpp"
-  "os/linux/AudioOutputPulse.cpp"
-  "os/linux/AudioInputPulse.cpp"
-  "os/linux/PulseAudioLoader.cpp"
   "os/posix/NetworkSocketPosix.cpp"
   "webrtc_dsp/webrtc/base/checks.cc"
   "webrtc_dsp/webrtc/base/stringutils.cc"
@@ -124,9 +120,6 @@
   "NetworkSocket.h"
   "os/linux/AudioInputALSA.h"
   "os/linux/AudioOutputALSA.h"
-  "os/linux/AudioOutputPulse.h"
-  "os/linux/AudioInputPulse.h"
-  "os/linux/PulseAudioLoader.h"
   "os/posix/NetworkSocketPosix.h"
   "webrtc_dsp/webrtc/base/array_view.h"
   "webrtc_dsp/webrtc/base/atomicops.h"
