--- cmd/syncthing/main.go	2014-06-13 05:29:07.566946128 +0700
+++ cmd/syncthing/main.go	2014-06-13 05:32:31.603821890 +0700
@@ -110,11 +110,9 @@
 func main() {
 	var reset bool
 	var showVersion bool
-	var doUpgrade bool
 	flag.StringVar(&confDir, "home", getDefaultConfDir(), "Set configuration directory")
 	flag.BoolVar(&reset, "reset", false, "Prepare to resync from cluster")
 	flag.BoolVar(&showVersion, "version", false, "Show version")
-	flag.BoolVar(&doUpgrade, "upgrade", false, "Perform upgrade")
 	flag.IntVar(&logFlags, "logflags", logFlags, "Set log flags")
 	flag.Usage = usageFor(flag.CommandLine, usage, extraUsage)
 	flag.Parse()
@@ -126,14 +124,6 @@
 
 	l.SetFlags(logFlags)
 
-	if doUpgrade {
-		err := upgrade()
-		if err != nil {
-			l.Fatalln(err)
-		}
-		return
-	}
-
 	if len(os.Getenv("GOGC")) == 0 {
 		debug.SetGCPercent(25)
 	}
