diff --git a/ngx_http_upstream_check_module.c b/ngx_http_upstream_check_module.c
index b628436..20e821e 100644
--- a/ngx_http_upstream_check_module.c
+++ b/ngx_http_upstream_check_module.c
@@ -6,6 +6,7 @@
 
 
 #include <nginx.h>
+#include <ngx_stream.h>
 #include "ngx_http_upstream_check_module.h"
 
 
@@ -494,6 +494,55 @@ static ngx_conf_bitmask_t  ngx_check_http_expect_alive_masks[] = {
 static ngx_command_t  ngx_http_upstream_check_commands[] = {
 
     { ngx_string("check"),
+      NGX_STREAM_UPS_CONF|NGX_CONF_1MORE,
+      ngx_http_upstream_check,
+      0,
+      0,
+      NULL },
+
+    { ngx_string("check_keepalive_requests"),
+      NGX_STREAM_UPS_CONF|NGX_CONF_TAKE1,
+      ngx_http_upstream_check_keepalive_requests,
+      0,
+      0,
+      NULL },
+
+    { ngx_string("check_http_send"),
+      NGX_STREAM_UPS_CONF|NGX_CONF_TAKE1,
+      ngx_http_upstream_check_http_send,
+      0,
+      0,
+      NULL },
+
+    { ngx_string("check_http_expect_alive"),
+      NGX_STREAM_UPS_CONF|NGX_CONF_1MORE,
+      ngx_http_upstream_check_http_expect_alive,
+      0,
+      0,
+      NULL },
+
+    { ngx_string("check_fastcgi_param"),
+      NGX_STREAM_UPS_CONF|NGX_CONF_TAKE2,
+      ngx_http_upstream_check_fastcgi_params,
+      0,
+      0,
+      NULL },
+
+    { ngx_string("check_shm_size"),
+      NGX_HTTP_MAIN_CONF|NGX_CONF_TAKE1,
+      ngx_http_upstream_check_shm_size,
+      0,
+      0,
+      NULL },
+
+    { ngx_string("check_status"),
+      NGX_HTTP_SRV_CONF|NGX_HTTP_LOC_CONF|NGX_CONF_TAKE1|NGX_CONF_NOARGS,
+      ngx_http_upstream_check_status,
+      0,
+      0,
+      NULL },
+
+    { ngx_string("check"),
       NGX_HTTP_UPS_CONF|NGX_CONF_1MORE,
       ngx_http_upstream_check,
       0,
