diff --git a/Telegram/SourceFiles/platform/linux/file_utilities_linux.cpp b/Telegram/SourceFiles/platform/linux/file_utilities_linux.cpp
index 4cb9a9065..f26f20a5c 100644
--- a/Telegram/SourceFiles/platform/linux/file_utilities_linux.cpp
+++ b/Telegram/SourceFiles/platform/linux/file_utilities_linux.cpp
@@ -94,6 +94,7 @@ void UnsafeShowInFolder(const QString &filepath) {
 
 namespace FileDialog {
 namespace {
+#ifndef TDESKTOP_DISABLE_GTK_INTEGRATION
 
 // GTK file chooser image preview: thanks to Chromium
 
@@ -104,9 +105,11 @@ namespace {
 // be preserved.
 constexpr auto kPreviewWidth = 256;
 constexpr auto kPreviewHeight = 512;
+#endif // !TDESKTOP_DISABLE_GTK_INTEGRATION
 
 using Type = ::FileDialog::internal::Type;
 
+#ifndef TDESKTOP_DISABLE_GTK_INTEGRATION
 bool NativeSupported() {
 	return Platform::internal::GdkHelperLoaded()
 		&& (Libs::gtk_widget_hide_on_delete != nullptr)
@@ -190,16 +193,20 @@ bool GetNative(QStringList &files, QByteArray &remoteContent, const QString &cap
 	remoteContent = QByteArray();
 	return false;
 }
+#endif // !TDESKTOP_DISABLE_GTK_INTEGRATION
 
 } // namespace
 
 bool Get(QStringList &files, QByteArray &remoteContent, const QString &caption, const QString &filter, Type type, QString startFile) {
+#ifndef TDESKTOP_DISABLE_GTK_INTEGRATION
 	if (NativeSupported()) {
 		return GetNative(files, remoteContent, caption, filter, type, startFile);
 	}
+#endif // !TDESKTOP_DISABLE_GTK_INTEGRATION
 	return ::FileDialog::internal::GetDefault(files, remoteContent, caption, filter, type, startFile);
 }
 
+#ifndef TDESKTOP_DISABLE_GTK_INTEGRATION
 namespace internal {
 
 QGtkDialog::QGtkDialog(GtkWidget *gtkWidget) : gtkWidget(gtkWidget) {
@@ -301,6 +308,7 @@ void QGtkDialog::onParentWindowDestroyed() {
 	// The Gtk*DialogHelper classes own this object. Make sure the parent doesn't delete it.
 	setParent(nullptr);
 }
+#endif // !TDESKTOP_DISABLE_GTK_INTEGRATION
 
 namespace {
 
@@ -320,6 +328,7 @@ QStringList cleanFilterList(const QString &filter) {
 
 } // namespace
 
+#ifndef TDESKTOP_DISABLE_GTK_INTEGRATION
 GtkFileDialog::GtkFileDialog(QWidget *parent, const QString &caption, const QString &directory, const QString &filter) : QDialog(parent)
 , _windowTitle(caption)
 , _initialDirectory(directory) {
@@ -623,5 +632,6 @@ void GtkFileDialog::setNameFilters(const QStringList &filters) {
 }
 
 } // namespace internal
+#endif // !TDESKTOP_DISABLE_GTK_INTEGRATION
 } // namespace FileDialog
 } // namespace Platform
